pages/building-packages-for-runtimes.md

```diff
@@ -0,0 +1,239 @@
+
+One example is, by trying load `react-router@5.2.0` into the browser as `esm` module. Fails to load it, as the `entry` uses `process.env.NODE_ENV`
+
+[https://unpkg.com/browse/react-router@5.2.0/index.js](https://unpkg.com/browse/react-router@5.2.0/index.js)
+
+```jsx
+// react-router/index.js
+if (process.env.NODE_ENV === "production") {
+  module.exports = require("./cjs/react-router.min.js");
+} else {
+  module.exports = require("./cjs/react-router.js");
+}
+```
+
+This is handled, if we are importing the process from `node:process` . Then bundlers and CDN’s which build these packages will be able to detect these built-ins usage. And polyfill them accordingly.
+
+So, the above snippet can become
+
+```jsx
+import process from "node:process";
+
+if (process.env.NODE_ENV === "production") {
+  module.exports = require("./cjs/react-router.min.js");
+} else {
+  module.exports = require("./cjs/react-router.js");
+}
+```
+
+This kinda of context and env aware of usage of built-ins. Will give better control for the bundlers to polyfill only the modules that are actually needed. Instead of polyfilling all the node built-ins every time a build is carried out for browsers.
+
+## JSPM
+
+We have seen multiple use cases for improving package compatibility within the ecosystem. Now, let's examine how `jspm` handles pre-existing packages in npm. It’s a well known issue when esm modules went into stable. The transition was not at all smooth in the eco-system. Because, for a project to work in js eco-system now.
+
+The build pipelines became so complex, it’s a proper combination of `bundlers`, `transpilers`, `node` and multiple module `formats`. Loading a `cjs` module from `npm` into project that is configuring to build as `esm` and vice-versa became so complex. At some point left most of the systems out of sync.
+
+`JSPM` builds all packages from `npm` ahead of time to spec complaint esm modules. And serves them using a distributed global CDN, regardless of the format in which the packages are published to npm. This makes it seamless for loading any module from `npm` and use it in any project at anytime.
+
+**Fields inside package.json**
+
+**browser**
+
+If a package is specifying [browser](https://docs.npmjs.com/cli/v9/configuring-npm/package-json#browser) field in the package.json. And when these packages are used in generating `import-maps` for browser as target. Then jspm uses this field instead of `main`, as the package authors explicitly stating the module that needs to be used when loading the same build inside a browser environments.
+
+**module**
+
+Module build detection in jspm. Let’s say, we are exposing `cjs` build from `main` and `esm` version in `module` field. And we don’t have `type` as `module` set on the package. Nor, the entry point is ending with `.mjs` extension. Something like below
+
+```json
+"main": "dist/index.js",
```

Comment:
The main field is also no longer needed - we can just use `"exports"`.

For example:

```json
"exports": {
  "import": "./dist/index.es.js",
  "default": "./dist/index.js"
}
```

works just fine now in all environments.